axelrod/strategies/axelrod_second.py

```diff
@@ -1454,3 +1454,50 @@ def strategy(self, opponent: Player) -> Action:
         self.flack *= 0.5  # Defections have half-life of one round
 
         return random_choice(1.0 - self.flack)
+
+
+class Leyvraz(Player):
+    """
+    Strategy submitted to Axelrod's second tournament by Fransois Leyvraz
+    (K68R) and came in eleventh in that tournament.
+
+    The strategy uses the opponent's last three moves to decide on an action
+    based on the following ordered rules.
+
+    1. If opponent Defected last two turns, then Defect with prob 75%.
+    2. If opponent Defected three turns ago, then Cooperate.
+    3. If opponent Defected two turns ago, then Defect.
+    4. If opponent Defected last turn, then Defect with prob 50%.
+    5. Otherwise (all Coopertaions), then Cooperate.
+
+    Names:
+
+    - Leyvraz: [Axelrod1980b]_
+    """
+
+    name = 'Leyvraz'
+    classifier = {
+        'memory_depth': 3,
+        'stochastic': True,
+        'makes_use_of': set(),
+        'long_run_time': False,
+        'inspects_source': False,
+        'manipulates_source': False,
+        'manipulates_state': False
+    }
+
+    def strategy(self, opponent: Player) -> Action:
+        recent_history = [C, C, C]  # Default to C.
+        for go_back in range(1, 4):
+            if len(opponent.history) >= go_back:
+                recent_history[-go_back] = opponent.history[-go_back]
+
+        if recent_history[-1] == D and recent_history[-2] == D:
+            return random_choice(0.25)
+        if recent_history[-3] == D:
+            return C
+        if recent_history[-2] == D:
+            return D
+        if recent_history[-1] == D:
+            return random_choice(0.5)
+        return C  # recent_history == [C, C, C]
```

Comment:
Just a suggestion to consider rather than a request, but might this be easier to read with a dict mapping recent_history to action? Something like:

```
history_to_action = {
    [C, C, C]: C,
    [C, C, D]: random_choice(0.5)
etc....
}

return  history_to_action[recent_history]
```

Reply:
I think this actually is not ideal in this case (unless I'm misunderstanding):

```
[C, C, D]: random_choice(0.5)
```

Implies `history_to_action` would need to be redefined at every call of `strategy` to ensure the random sample is different which in turn means we'd be using the random module even when not needed (which bumps the seed and makes the strategy slower).

Reply:
One way around it (if you did want to use a `dict`) would be to use the dict to map to probabilities:

{
    [C, C, C]: 1,
    [C, C, D]: 0.1
}

and then use `random_choice(history_to_action[recent_history])` (the `random_choice` is smart so if passed 0 or 1 doesn't sample).

Reply:
I think @drvinceknight is right here, but I'll keep in mind.

Reply:
I went ahead and implemented this.  I agree that it's a little more readable.